### PR TITLE
Authorization/Permissions Prompt studio

### PR DESCRIPTION
**Pull request by johnyrahul**
## What

-  https://zipstack.atlassian.net/wiki/spaces/PT/pages/334233613/Entity+Resource+Prompt+Studio+Tool+Authorization
-  Re-structured prompt-studio sub apps .
- Organized API/ Endpoint exposure to work seamlessly with Django permissions

## Why

- Today we don't have any option to collaborate on prompt studio projects.

## How

- Using the Django permission model.

## Database Migrations

-  Introduced Many to Many linking with
-  User and Prompt Studio Project
-  User and Exported tool
-  Added a new boolean field in the exported tool to specify if it shared with the entire org

## Env Config

- 

## Relevant Docs

-

## Related Issues or PRs

- https://github.com/Zipstack/unstract/pull/117

## Dependencies Versions

-

## Notes on Testing

-

## Screenshots
image
image
image

## Checklist

I have read and understood the [Contribution Guidelines]().

### CODE REVIEW

**Comment by johnyrahul** [comment]
@jagadeeswaran-zipstack  Please take look into the sonar issue, most of them should be addressable with small changes

**Comment by johnyrahul** [comment]
> @jagadeeswaran-zipstack Please take look into the sonar issue, most of them should be addressable with small changes

Thanks @mohamed-siddhiq  for closing the sonar related fixes.

File: backend/prompt_studio/prompt_studio_core/models.py [hunk 1]
```diff
@@ -101,3 +116,11 @@ class CustomTool(BaseModel):
     enable_challenge = models.BooleanField(
         db_comment="Flag to enable or disable challenge", default=False
     )
+
+    # Introduced field to establish M2M relation between users and custom_tool.
+    # This will introduce intermediary table which relates both the models.
+    shared_users = models.ManyToManyField(
```
Comment (harini-venkataraman) [inline backend/prompt_studio/prompt_studio_core/models.py:122]: @johnyrahul What will the value of `shared_users` be for Prompt projects created before this field was introduced?

Comment (johnyrahul) [reply backend/prompt_studio/prompt_studio_core/models.py:122]: @harini-venkataraman  when we introduce this relation it creates an intermediary table.

Attached screenshot for better understanding
image

File: backend/prompt_studio/prompt_studio_registry/models.py [hunk 2]
```diff
@@ -72,3 +86,14 @@ class PromptStudioRegistry(BaseModel):
         blank=True,
         editable=False,
     )
+    shared_to_org = models.BooleanField(
```
Comment (harini-venkataraman) [inline backend/prompt_studio/prompt_studio_registry/models.py:89]: @johnyrahul [DOUBT] Does the intermediary table introduced take care of cascade delete here?

Comment (johnyrahul) [reply backend/prompt_studio/prompt_studio_registry/models.py:89]: No @harini-venkataraman  we have handle it. [Refer](https://docs.djangoproject.com/en/2.0/ref/models/relations/#django.db.models.fields.related.RelatedManager.clear)
We have handled this [here](https://github.com/Zipstack/unstract/blob/93eac974ef916bc4151aff34776f98ed7236c721/backend/account/authentication_controller.py#L406)

File: backend/prompt_studio/prompt_studio_core/views.py [hunk 3]
```diff
@@ -44,29 +69,30 @@ class PromptStudioCoreView(viewsets.ModelViewSet):
...
+
     def get_queryset(self) -> Optional[QuerySet]:
         filter_args = FilterHelper.build_filter_args(
             self.request,
             ToolStudioKeys.CREATED_BY,
         )
         if filter_args:
-            queryset = CustomTool.objects.filter(
-                created_by=self.request.user, **filter_args
+            queryset = CustomTool.objects.for_user(self.request.user).filter(
```
Comment (chandrasekharan-zipstack) [inline backend/prompt_studio/prompt_studio_core/views.py:88]: @johnyrahul we could simplify it by adding this as a method to the model [like so](https://stackoverflow.com/a/59649302/12864260)

Comment (johnyrahul) [reply backend/prompt_studio/prompt_studio_core/views.py:88]: Confused!.. That is what I have done here. @chandrasekharan-zipstack 

I think the if else is not required will remove them

Comment (chandrasekharan-zipstack) [reply backend/prompt_studio/prompt_studio_core/views.py]: I think the else is okay here, it'll not be needed if we specify a default `queryset` as a class instance

**Review by vishnuszipstack — APPROVED** [review]
LGTM

File: frontend/src/components/custom-tools/document-manager/DocumentManager.jsx [hunk 4]
```diff
@@ -127,7 +127,7 @@ function DocumentManager({ generateIndex, handleUpdateTool, handleDocChange }) {
 
     const requestOptions = {
       method: "GET",
-      url: `/api/v1/unstract/${sessionDetails?.orgId}/prompt-studio/file/fetch_contents?document_id=${selectedDoc?.document_id}&view_type=${viewType}&tool_id=${details?.tool_id}`,
+      url: `/api/v1/unstract/${sessionDetails?.orgId}/prompt-studio/file/${details?.tool_id}?document_id=${selectedDoc?.document_id}&view_type=${viewType}`,
```
Comment (tahierhussain) [inline frontend/src/components/custom-tools/document-manager/DocumentManager.jsx:130]: @johnyrahul This API is also being called inside unstract-cloud, where we get the summary content. Can you please add the change [there](https://github.com/Zipstack/unstract-cloud/blob/5adbeb24ba768a494fe4900553fd035c354930e5/frontend/src/plugins/summarize-view/SummarizeView.jsx#L40) as well?

**Pull request merged**